File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        // shortest bath problem between beginWord and endWord
        // each word is a node in graph 
        // word a can transofro to word b => a -> b
        // build the graph with 0: beginword 1 - n word in wordList 
        int n = wordList.size();
        unordered_set<string> words(wordList.begin(), wordList.end());
        if (!words.count(endWord)) {
            return 0;
        }
        unordered_set<string> visited;
        queue<pair<string, int>> q;

        q.emplace(beginWord, 1);
        while (!q.empty()) {
            auto [word, step] = q.front();
            q.pop();

            if (visited.count(word)) {
                continue;
            }
            visited.insert(word);

            if (endWord == word) {
                return step;
            }

            for (int i = 0; i < word.size(); ++i) {
                auto origin = word[i];

                for (char chr = 'a'; chr <= 'z'; chr++) {
                    if (origin == chr) {
                        continue;
                    }
                    word[i] = chr;
                    if (!visited.count(word) && words.count(word)) {
                        words.erase(word);
                        q.emplace(word, step + 1);
                    }
                }
                word[i] = origin;
            }
        }
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        # bfs start from beginword
        # node: word in wordlist or beginword or endword
        # edge: node a -> node b ifnode b can be obtain by change only one letter in node a
        # if reach endWord update length(initialize 0)
        # time complexity o(V + E) = o(n + )
        # time complexity o (n * 26 * len of a word) x
        # we need ot find min word so we need to backtrack
        # backtrack so time complexity o(n^len of list)?
        # dfs may be quick

        # in bfs

        #     for i in range(word length):
        #         for char in 26 char
        #             change the word[i] to char
        words = set(wordList)
        visited = {beginWord}
        queue = deque([(beginWord, 1)])

        while queue:
            cur, cur_num_trans = queue.popleft()

            if cur == endWord:
                return cur_num_trans

            for i in range(len(cur)):
                cur_chr = cur[i]
                for j in range(26):
                    nbr = list(cur)
                    nbr[i] = chr(ord('a') + j)
                    nbr = "".join(nbr)

                    if nbr not in visited and nbr[i] != cur_chr and nbr in words:
                        visited.add(nbr)
                        queue.append((nbr, cur_num_trans + 1))
        
        return 0 
                

        
        